lms/djangoapps/course_api/api.py

```diff
@@ -170,19 +172,38 @@ def list_course_keys(request, username, role):
     """
     user = get_effective_user(request.user, username)
 
-    course_keys = CourseOverview.get_all_course_keys()
+    all_course_keys = CourseOverview.get_all_course_keys()
 
     # Global staff have access to all courses. Filter courses for non-global staff.
     if GlobalStaff().has_user(user):
-        return course_keys
-
-    return LazySequence(
-        (
-            course_key for course_key in course_keys
-            if has_access(user, role, course_key)
-        ),
-        est_len=len(course_keys)
-    )
+        return all_course_keys
+
+    if role == 'staff':
+        # This short-circuit implementation bypasses has_access() which we think is too slow for some users when
+        # evaluating staff-level course access for Insights.  Various tickets have context on this issue: CR-2487,
+        # TNL-7448, DESUPPORT-416, and probably more.
+        #
+        # This is a simplified implementation that does not consider org-level access grants (e.g. when course_id is
+        # blank).
+        filtered_course_keys = (
+            CourseAccessRole.objects.filter(
```

Comment:
Okay, so this actually returns Libraries as well, which is probably not intended. Maybe join against CourseOverview?

Reply:
So, I couldn't figure out a way to actually do a join without injecting raw SQL (since there's no ForeignKey field that would allow me to do a join via typical ORM APIs), so this solution involves a subquery.

Reply:
You can do it with [extra](https://docs.djangoproject.com/en/2.2/ref/models/querysets/#django.db.models.query.QuerySet.extra), like we ended up doing for certs: https://github.com/edx/edx-platform/pull/20547/files

But I'm fine with this too for now.

Reply:
I'm tempted to change it to use extra because I don't want to second-guess this change if performance issues continue to crop up.  I would expect the subquery version would perform almost proportionally to the number of courses returned, which for the problematic users can be 1000+

Reply:
With extra(), the new SQL is
```
SELECT DISTINCT "student_courseaccessrole"."course_id"
FROM "student_courseaccessrole" , "course_overviews_courseoverview"
WHERE
  ("student_courseaccessrole"."role" IN (staff, instructor)
  AND "student_courseaccessrole"."user_id" = 3 AND (course_id = course_overviews_courseoverview.id)
  AND NOT ("student_courseaccessrole"."course_id" = ))
ORDER BY "student_courseaccessrole"."course_id" ASC
```

Reply:
(We should maybe double-check the libraries tab loads as expected in Studio Home at that might be the one place the Libraries check matters? @ormsbee ?)